pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -803,7 +824,7 @@ void ackReceived(ClientCnx cnx, long sequenceId, long ledgerId, long entryId) {
         if (callback) {
             op = pendingCallbacks.poll();
             if (op != null) {
-                lastSequenceIdPublished = op.sequenceId + op.numMessagesInBatch - 1;
+                lastSequenceIdPublished = getHighestSequenceId(op);
```

Comment:
this should :

```
lastSequenceIdPublished = Math.max(lastSequenceIdPublished, getHighestSequenceId(op));
```

because the duplicated publishes will still succeed, right. It can override a larger sequence Id with a smaller one.